conjure-http/src/client.rs

```diff
@@ -0,0 +1,76 @@
+// Copyright 2019 Palantir Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! The Conjure HTTP client API.
+
+use conjure_error::Error;
+use http::{Request, Response};
+use std::io::{Cursor, Read, Write};
+
+/// A trait implemented by HTTP client implementations.
+pub trait Client {
+    /// The client's response body type.
+    type ResponseBody: Read;
+
+    /// Makes an HTTP request.
+    ///
+    /// The request's URI will be absolute-form, and it is the responsibility of the client to add the authority and
+    /// any extra context path required. The request body will be unencoded, and the request will not include a
+    /// `Content-Length` header.
+    ///
+    /// A response must only be returned if it has a 2xx status code. The client is responsible for handling all other
```

Comment:
not sure if worth mentioning 204s here

Reply:
I don't think the client needs to do anything special there.

Reply:
hm, so there's some specific stuff in the wire spec around 204s - but not sure how much validation should happen on the client side vs server side for that.

If result type is optional<T> and server returns empty 200, that goes against
> the de-aliased return type is optional<T> and the value is not present: servers must send 204 No Content.

but not sure if the client needs to care

https://palantir.github.io/conjure/#/docs/spec/wire?id=_31-status-codes

Reply:
The client itself doesn't have the context to know if a 204 is ok or not. It is responsible for knowing that the HTTP response is well formed (i.e. a 204 doesn't have a body), but IMO that's implied by it being an HTTP client in the first place.